### PR TITLE
add mongodbQueue implementation

### PR DESCRIPTION
**Pull request by wei-lee**
## Motive

In order to scale the sync framework, we need a mongodb-backed queue implementation.

## Changes

Add a queue implementation that is backed by mongodb. It's a thin wrapper around the `mongodb-queue` module.

## JIRA

https://issues.jboss.org/browse/FH-3139

ping @david-martin

### CODE REVIEW

File: lib/sync/mongodbQueue.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/**
+ * construct a new queue that is backed by mongodb.
+ * @param {String} name the name of the queue.
+ * @param {Object} opts configuration options for the queue
+ * @param {Object} opts.mongodb an instance of the mongodb connection. This is required.
+ * @param {Number} opts.visibility see https://github.com/chilts/mongodb-queue#visibility---message-visibility-window
+ */
+function MongodbQueue(name, opts) {
```
Comment (fh-qube) [inline lib/sync/mongodbQueue.js]: INFO Rename this 'MongodbQueue' function to match the regular expression ^[a-z][a-zA-Z0-9]*$ [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS100)

File: lib/sync/mongodbQueue.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }
+  return cb(null, self.queue);
+};
+
+/**
+ * See https://github.com/chilts/mongodb-queue for details about those methods.
+ * We may not need all those methods, we can review them and remove the ones that we don't need.
+ */
+['add', 'get', 'ack', 'ping', 'total', 'size', 'inFlight', 'done', 'clean'].forEach(function(methodName){
+  MongodbQueue.prototype[methodName] = function() {
```
Comment (fh-qube) [inline lib/sync/mongodbQueue.js]: BLOCKER Add a semicolon at the end of this statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)

**Comment by fh-qube** [comment]
SonarQube analysis reported 2 issues:
* BLOCKER 1 blocker
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by fh-qube** [comment]
SonarQube analysis reported 1 issue:
* INFO 1 info

Watch the comments in this conversation to review them.

File: lib/sync/mongodbQueue.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
+var mongodbQ = require('mongodb-queue');
```
Comment (david-martin) [inline lib/sync/mongodbQueue.js]: @wei-lee you can probably use this in sync/index.js now. There's placeholders for where queues are created directly with mongodb-queue already https://github.com/feedhenry/fh-mbaas-api/blob/FH-3123-sync-horizontal-scaling/lib/sync/index.js#L121-L125

**Comment by evanshortiss** [comment]
@david-martin @wei-lee would strongly suggest investigating this module by @MIkeyBurkman https://github.com/MikeyBurkman/mongo-queue

We're using it in production and it's fantastic.

**Comment by MikeyBurkman** [comment]
@evanshortiss So from reading the Jira, my mongo queue currently doesn't quite match their requirements, as mine is not multi-thread safe, and the Jira states that it will have multiple processes hitting it. I've thought about making it thread safe (basically doing what this mongodb-queue module is doing, with doing an atomic findAndModify() one record at a time, to prevent multiple processes from processing a record), but we haven't needed it for TKE so I haven't spent any time on it.

It is still rather interesting to note the differences between my module and mongodb-queue module. Mine is much more configure-and-forget, while mongodb-queue requires you to, for instance, manually get records off the queue, process them, and acknowledge them when you're complete. (I'm not sure why they didn't just use a callback instead of an ack but whatever.) The differences between mine and his are actually quite numerable.

**Comment by fh-qube** [comment]
SonarQube analysis reported 14 issues:
* MAJOR 2 major
* MINOR 3 minor
* INFO 9 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Move the declaration of "invokeFunctions" before this usage.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L32) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L58) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L15) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L51) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L75) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L42) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L50) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L67) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L43) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/54ff9d2edc9ff222b33f37a51485c9129ecda5b4/lib/sync/index.js#L59) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* ... 3 more

**Comment by fh-qube** [comment]
SonarQube analysis reported 14 issues:
* MAJOR 2 major
* MINOR 3 minor
* INFO 9 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L59) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L67) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MAJOR [Add the "var" keyword to this declaration of "syncScheduler".](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L139) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2703)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L51) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L50) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L104) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MAJOR [Move the declaration of "invokeFunctions" before this usage.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L32) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L43) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L15) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/0129175c058bc3107280c5e43615b4761b78092c/lib/sync/index.js#L95) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* ... 3 more

**Comment by david-martin** [comment]
Merging as PR #50 will fix the CI failures

**Pull request merged**